src/main/kotlin/org/rust/lang/core/macros/MacroExpansion.kt

```diff
@@ -5,16 +5,103 @@
 
 package org.rust.lang.core.macros
 
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFileFactory
+import com.intellij.psi.util.PsiTreeUtil
+import org.rust.lang.RsFileType
+import org.rust.lang.core.psi.RsFile
 import org.rust.lang.core.psi.RsMacroCall
+import org.rust.lang.core.psi.RsMacroDefinition
+import org.rust.lang.core.psi.RsMacroDefinitionBody
 import org.rust.lang.core.psi.ext.RsElement
+import org.rust.lang.core.psi.ext.descendantOfTypeStrict
 import org.rust.lang.core.psi.ext.macroName
 
 fun expandMacro(call: RsMacroCall): List<ExpansionResult>? {
     val context = call.context as? RsElement ?: return null
     val result = when {
         call.macroName == "lazy_static" -> expandLazyStatic(call)?.let { listOf(it) }
-        else -> null
+        else -> {
+            val def = call.reference.resolve() as? RsMacroDefinition ?: return null
+            expandMacro(def, call)
+        }
     }
     result?.forEach { it.setContext(context) }
     return result
 }
+
+fun createMacroExpansionFile(project: Project, text: String): RsFile {
+    return PsiFileFactory.getInstance(project)
+        .createFileFromText("EXPANSION_MACRO.rs", RsFileType, text)
+        as RsFile
+}
+
+inline fun <reified I : RsElement> createInMemoryPsi(project: Project, code: String): I? {
+    return createMacroExpansionFile(project, code)
+        .descendantOfTypeStrict()
+}
+
+private val RsMacroDefinition.body: RsMacroDefinitionBody?
+    get() {
+        val stub = stub ?: return macroDefinitionBody
+        val text = stub.macroBody ?: return null
```

Comment:
Hmm, can you please explain why we storing a raw text to stub instead of stubbing all nodes? Is it something fundamental or it's just easier to implement?

Reply:
And we even don't cache expanded body anywhere and should parse it again for every usage :(

Reply:
It's much easier to implement (it's hard to stub leaf elements, and we need them here) and, I think, might actually be more efficient, because we wan't be storing the actual tree structure.

Reply:
Let's try to insert a cache here. I wonder how this will impact performance.